extensions/modules/src/org/exist/xquery/modules/expathrepo/InstallFunction.java

```diff
@@ -150,15 +149,23 @@ private URI _getURI(String s) throws XPathException
         }
     }
 
-    private BinaryDocument _getDocument(String path) throws XPathException {
-    	try {
-			XmldbURI uri = XmldbURI.createInternal(path);
-			DocumentImpl doc = context.getBroker().getXMLResource(uri, LockMode.READ_LOCK);
-			if (doc.getResourceType() != DocumentImpl.BINARY_FILE)
-				throw new XPathException(this, EXPathErrorCode.EXPDY001, path + " is not a valid .xar", new StringValue(path));
-			return (BinaryDocument) doc;
-		} catch (PermissionDeniedException e) {
-			throw new XPathException(this, EXPathErrorCode.EXPDY003, e.getMessage(), new StringValue(path), e);
-		}
+  private BinaryDocument _getDocument(String path) throws XPathException {
+    try {
+      XmldbURI uri = XmldbURI.createInternal(path);
+      DocumentImpl doc = context.getBroker().getXMLResource(uri, LockMode.READ_LOCK);
+      if (doc == null) {
+        notValidXar(path);
+      } else if (doc.getResourceType() != DocumentImpl.BINARY_FILE) {
+        doc.getUpdateLock().release(LockMode.READ_LOCK);
+        notValidXar(path);
+      }
+      return (BinaryDocument) doc;
+    } catch (PermissionDeniedException e) {
+      throw new XPathException(this, EXPathErrorCode.EXPDY003, e.getMessage(), new StringValue(path), e);
     }
+  }
+
+  private void notValidXar(String path) throws XPathException {
+    throw new XPathException(this, EXPathErrorCode.EXPDY001, path + " is not a valid .xar", new StringValue(path));
```

Comment:
I'd suggest to have two simple throw exceptions on line 160 and 164 ; with this you could probably change the text into "Resource not found" and "Resource is not binary" .... which helps in finding a root cause.